### PR TITLE
dcp-386 project search by keywords

### PR DESCRIPTION
**Pull request by amnonkhen**
related to ebi-ait/dcp-ingest-central#386

### CODE REVIEW

File: client/src/app/all-projects/all-projects.component.ts [hunk 1]
```diff
@@ -26,13 +27,15 @@ export class AllProjectsComponent implements OnInit, OnDestroy {
     ProjectColumn.wranglingState
   ];
   isWrangler: Boolean = true;
+  organ: '';
```
Comment (amnonkhen) [inline client/src/app/all-projects/all-projects.component.ts:30]:
```suggestion

```

File: client/src/app/all-projects/all-projects.component.html [hunk 2]
```diff
@@ -16,13 +16,24 @@ <h1 class="vf-text vf-text-heading--2">All Projects</h1>
     </button>
   </mat-form-field>
 
+  <mat-form-field>
+    <mat-select placeholder="Search Type"
+                [(value)]="searchType"
+                (selectionChange)="onChangeSearchType($event)">
+      <mat-option value="AllKeywords">All Keywords</mat-option>
+      <mat-option value="AnyKeyword">Any Keyword</mat-option>
+      <mat-option value="ExactMatch">Exact Match</mat-option>
```
Comment (aaclan-ebi) [inline client/src/app/all-projects/all-projects.component.html:25]: noice!

File: client/src/app/shared/data-sources/project-data-source.spec.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+describe('ProjectDataSource', () => {
+
+  beforeEach(() => {
+  });
+
+  describe('filterByFieldAndValue', () => {
+    it('filterByFieldAndValue', (done) => {
+      const ds = new ProjectDataSource(null);
+      ds.filterByFieldAndValue('x', 'y');
```
Comment (aaclan-ebi) [inline client/src/app/shared/data-sources/project-data-source.spec.ts:15]: should there be an assertion here?

**Review by aaclan-ebi — APPROVED** [review]

File: client/src/app/shared/data-sources/project-data-source.spec.ts (hunk 3 above)
Comment (amnonkhen) [reply client/src/app/shared/data-sources/project-data-source.spec.ts:15]: There will be, I am thinking about what should be tested still. I wanted to get some code out for review.

Comment (aaclan-ebi) [reply client/src/app/shared/data-sources/project-data-source.spec.ts:15]: ah, I see that you're excluding it for now.

**Review by aaclan-ebi — DISMISSED** [review]

File: client/src/app/shared/data-sources/project-data-source.spec.ts (hunk 3 above)
Comment (amnonkhen) [reply client/src/app/shared/data-sources/project-data-source.spec.ts:15]: I am still thinking about what to test here.

File: client/src/app/shared/data-sources/project-data-source.ts [hunk 4]
```diff
@@ -5,37 +5,43 @@ import {Project} from '../models/project';
...
 
-    public search(searchString: string): void {
-        const queryData = { ...this.getQueryData(), search: searchString, page: 0 };
-        if (searchString === '') {
-            delete queryData['search'];
-        }
-        this.setQueryData(queryData);
-    }
+  public filterByWranglingState(wranglingState: string): void {
+    this.filterByFieldAndValue('wranglingState', wranglingState);
```
Comment (jacobwindsor) [inline client/src/app/shared/data-sources/project-data-source.ts:31]: This is nice but I would use partial application here. So e.g.:

```
const filterByFieldAndValue = (fieldName) => (fieldValue) => { ... }

const filterByWrangler = this.filterByFieldAndValue('wrangler')
```

**Review by jacobwindsor — CHANGES_REQUESTED** [review]

File: client/src/app/shared/data-sources/project-data-source.ts (hunk 4 above)
Comment (jacobwindsor) [reply client/src/app/shared/data-sources/project-data-source.ts:31]: Oh and define a `filterBySearchType` public method

**Review by jacobwindsor — CHANGES_REQUESTED** [review]

File: client/src/app/shared/data-sources/project-data-source.ts (hunk 4 above)
Comment (amnonkhen) [reply client/src/app/shared/data-sources/project-data-source.ts:31]: If find that my head explodes when I see  `x => f => z => ...`, but I will give it a go, and see how it looks.

**Review by jacobwindsor — APPROVED** [review]

File: client/src/app/all-projects/all-projects.component.spec.ts [hunk 5]
```diff
@@ -43,18 +54,46 @@ describe('AllProjectsComponent', () => {
...
+      .callThrough();
+    spyOn(component, 'getProjects').and.callThrough();
+    component.ngOnInit();
+
+    const select = fixture.debugElement.query(By.css('.search-type'));
+    select.triggerEventHandler('selectionChange', { source: null, value: 'AllKeywords' });
+    expect(component.onChangeSearchType).toHaveBeenCalled();
+
+    setTimeout(() => {
+      // Messy solution: Works by moving this expectation to the task queue so that it is executed after everything
```
Comment (aaclan-ebi) [inline client/src/app/all-projects/all-projects.component.spec.ts:73]: still better than a timeout!

File: client/src/app/all-projects/all-projects.component.spec.ts [hunk 6]
```diff
@@ -43,18 +54,46 @@ describe('AllProjectsComponent', () => {
...
-    expect(component.searchText).toBe('');
-    expect(fixture.nativeElement.querySelector('input').value).toBe('');
+    expect(component.searchText)
+      .toBe('');
+    expect(fixture.nativeElement.querySelector('input').value)
+      .toBe('');
   });
 });
 
 class MockIngestService {
```
[the remark this answers is not in the document]
Comment (ke4) [reply client/src/app/all-projects/all-projects.component.spec.ts:92]: Yes, in line 19 its usage has been replace by a spy object.

File: client/src/app/all-projects/all-projects.component.ts [hunk 7]
```diff
@@ -43,7 +45,7 @@ export class AllProjectsComponent implements OnInit, OnDestroy {
   ngOnInit() {
     this.dataSource = new ProjectDataSource(this.getProjects.bind(this));
     this.dataSource.sortBy('updateDate', 'desc');
-    this.dataSource.connect(true).subscribe({
+    this.dataSource.connect(true, 30000).subscribe({
```
Comment (ke4) [inline client/src/app/all-projects/all-projects.component.ts:48]: Why is that magic number? Could we at least name it please?

File: client/src/app/my-projects/my-projects.component.ts [hunk 8]
```diff
@@ -33,7 +33,7 @@ export class MyProjectsComponent implements OnInit, OnDestroy {
     // protected against null user by the user-is-logged-in guard
     this.accountDataSource =  new SimpleDataSource(this.ingestService.getUserAccount.bind(this.ingestService));
     this.projectsDataSource = new PaginatedDataSource(params => {
-      return this.ingestService.getUserProjects({ ...params, sort: `${params.sort.column},${params.sort.direction}`})
+      return this.ingestService.getUserProjects({ ...params, sort: `${params.sort.column},${params.sort.direction}`});
```
Comment (ke4) [inline client/src/app/my-projects/my-projects.component.ts:36]: Semicolons in JavaScript are optional due to Automatic Semicolon Insertion (ASI)

**Review by ke4 — CHANGES_REQUESTED** [review]
Requested tiny changes, but overall looks fine. 👍  Tx

File: client/src/app/my-projects/my-projects.component.ts (hunk 8 above)
Comment (jacobwindsor) [reply client/src/app/my-projects/my-projects.component.ts:36]: It's a style thing. You should either use them or not and it seems that we use them

**Review by aaclan-ebi — APPROVED** [review]

**Review by aaclan-ebi — APPROVED** [review]

**Review by jacobwindsor — APPROVED** [review]

File: client/src/app/all-projects/all-projects.component.ts [hunk 9]
```diff
@@ -43,7 +47,7 @@ export class AllProjectsComponent implements OnInit, OnDestroy {
   ngOnInit() {
     this.dataSource = new ProjectDataSource(this.getProjects.bind(this));
     this.dataSource.sortBy('updateDate', 'desc');
-    this.dataSource.connect(true).subscribe({
+    this.dataSource.connect(true, THIRTY_SECONDS).subscribe({
```
Comment (ke4) [inline client/src/app/all-projects/all-projects.component.ts:50]: but what is THIRTY_SECONDS. Is it the value of the timeout?

Comment (amnonkhen) [reply client/src/app/all-projects/all-projects.component.ts:50]: Correct, could be improved.

**Review by ke4 — APPROVED** [review]
I added a small comment to variable naming, but I approved it

**Pull request merged**